module3/lessons/Week2/FeatureTesting.md

```diff
@@ -0,0 +1,327 @@
+* Set up a Test Database
+
+## Feature Testing
+
+### What are Feature Tests?
+
+* Feature tests mimic the behavior of the user: In the case of web apps, this behavior will be clicking, filling in forms, visiting new pages, etc. 
+* Just like a user, the feature test should not need to know about underlying code.
+* Based on user stories
+
+### What are User Stories?
+* A tool used to communicate user needs to software developers.
+* They are used in Agile Development, and it was first introduced in 1998 by proponents of Extreme Programming.
+* They describe what a user needs to do in order to fulfill a function.
+* They are part of our "top-down" design.
+
+```txt
+As a user
+When I visit the home page
+And I click on the link New Task
+Then I see a form to create a task
+
+And I fill in title
+And I fill in description
+And I click submit
+Then my task is saved
+```
+
+We can generalize this pattern as follows:
+
+```
+As a [user/user-type]
+When I [action]
+And I [action]
+And I [action]
+...
+Then [expected result]
+```
+
+Depending on how encompassing a user story is, you may want to break a single user story into multiple, smaller user stories.
+
+<section class="call-to-action">
+    <strong>Connecting to the Prework</strong><p>: What type of testing that you read about yesterday is most similar to this description of Feature Tests?</p>
+</section>
+
+## Creating a Test Project
+
+### The Components of a Web Application
+In order to create an environment where our test project has everything we need to create feature tests; we first need to identify what exists in our MVC application that allow us to interact with it via a client (the browser). 
+
+Any web application has two main features that need to exist in order to function:
+* **Server** - the server is the engine that allows the application to respond to requests.  The server starts up and _listens_ for requests.  When a request is received, the server sends back the appropriate response.
+* **Database** - the database allows user to interact with dynamic data.
+
+These components exist in our MVC application, and we need to create test versions of these two components in our test project.
+
+<section class="call-to-action">
+    <p>With a Partner - Review the <code>Program.cs</code> file in your starter project.</p>
+    <ul>
+        <li>Can you identify the code that sets up the server?</li>
+        <li>Can you identify the code that sets up the database?</li>
+        <li>Are there other files in the project that contribute to creating the server and database?</li>
+    </ul>
+</section>
+
+## Setting Up a Test Server Environment
+
+1. Add a new Project to your Solution - call the project `MvcMovie.FeatureTests`.
+2. Create a Project Reference from your test project to the `MvcMovie` project.
+3. Install the following packages:
+    * Microsoft.Extentions.Hosting
+    * Microsoft.AspNetCore.Hosting
+    * Microsoft.AspNetCore.Mvc.Testing (version 6.0.16)
```

Comment:
I originally wan't able to run my tests. It was because I had Microsoft.EntityFrameworkCore.Design installed with version 7. I downgraded to 6 and then it worked. I'm thinking maybe for all of the packages we install we should make sure students install version 6 not 7

Reply:
I updated to include all the versions that I used when installing.  I think that we will probably always run in to _some_ issues with versioning, but hopefully this helps!